Add API to Camera model to flip axes

jni opened the pull request:
In #7554 / #7488 we changed the default orientation of axes in 3D mode. This may be undesirable to some users, so this PR adds the ability to flip them.

It also turns out to be handy when displaying data which has specific absolute coordinates as well as a common convention for the display direction, such as geographical data. See the included example which both demonstrates how to use napari with xarray as well as how to use this new camera axis-flip API. Currently I'm pretty happy with the API (see below), but I'm very much open to alternative suggestions!

## API description

This adds an `orientation` attribute to our Camera model. The scene is assumed to have three axes, z, y, and x, which are aligned with the depth, vertical, and horizontal axes of the camera. Each of these axes can then point either:
- away or towards the camera
- down or up
- right or left

I used a StrEnum for each of these values, which makes the API both robust for type checking and easy to use, e.g.:

```python
viewer.camera.orientation = ('away', 'up', 'right')
```

I also added an `orientation2d` property, because it's strange to set three values to flip two axes, even though that is the Vispy model. (camera.flip is a 3-tuple even for 2D cameras in Vispy.) You can see its use in the example. To show latitude (-90 for south, 90 for north), we want the vertical axis to point up, not down like in the napari default. So we do:

```python
viewer.camera.orientation2d = ('up', 'right')
```

and the geographic raster of temperatures is displayed correctly:

![CleanShot 2025-03-03 at 12 42 18@2x](https://github.com/user-attachments/assets/03ee83dd-9272-4991-bf28-524ecd5ff321)

Without this API, the map ends up upside down:

![CleanShot 2025-03-04 at 18 29 50@2x](https://github.com/user-attachments/assets/bb51c197-9fe4-42c8-bd28-687309b46829)

jni committed:
Add orientation to the camera model

jni committed:
Add 2D orientation property and setter

jni committed:
Add example demonstrating display of xarray and flipping camera axis orientation

willingc commented on examples/xarray-latlon-timeseries.py:
```diff
@@ -0,0 +1,81 @@
+"""Displaying xarray data in napari.
+
+This example shows how to view xarray datasets in napari, including scale
+and translation information. Currently, napari cannot display
```
```suggestion
and translation information.

Currently, napari cannot display
```

willingc commented on napari/_vispy/camera.py:
```diff
@@ -3,6 +3,9 @@
 
 from napari._vispy.utils.quaternion import quaternion2euler_degrees
 
+VISPY_DEFAULT_ORIENTATION_2D = ('right', 'up', 'towards')  # xyz, not zyx
```
I would remove the not zyx or give a bit more context.

DragaDoncila replied:
Ok I just came to this line confused because further down we have `DEFAULT_ORIENTATION=('towards', 'down', 'right')` and in the example we have `('up', 'right')`, so I agree we need a more expressive comment here that our default axis ordering is flipped.

willingc commented on napari/_vispy/camera.py:
```diff
@@ -176,6 +181,21 @@ def _on_center_change(self):
         self.zoom = self._camera.zoom
 
+    def _on_orientation_change(self):
+        orientation_xyz = self._camera.orientation[::-1]
```
Is this reorienting from xyz to zyx? If so adding a 1 line docstring would give context.

willingc reviewed (commented):
A few questions. Looks cool.

jni committed:
Add h5netcdf to gallery dependencies (for xarray sample data)

jni committed:
Add comments to explain what _on_orientation_change is doing

jni committed:
Add paragraph break in example for clarity

jni committed:
Try to fix mypy

jni committed:
Change module docstring format to include title for gallery

willingc reviewed (approved):
Nice!

DragaDoncila commented on examples/xarray-latlon-timeseries.py:
```diff
@@ -0,0 +1,85 @@
+
+# latitude goes from -90 (south, down) to 90 (north, up),
+# so we make sure that the camera vertical axis points up.
+viewer.camera.orientation2d = ('up', 'right')
```
Would be great to actually add the image of the non-flipped camera into the description, to drive home the point!

jni replied:
what do you mean "add the image into the description?"

DragaDoncila replied:
Sorry, I meant, add the image loaded with default orientation to the PR description.

jni replied:
Done! The Upside Down lives!

DragaDoncila reviewed (approved):
Works as advertised on the box! I like the example too. Would be good to see a corresponding docs PR...

jni committed:
Add longer comment explaining why axis ordering is reversed in vispy camera

jni committed:
Add test for axis flip in 2D

jni committed:
Slightly simplify 2D orientation flip test

jni committed:
Add docstring for 2D orientation test

jni committed:
Add test for 3D camera orientation flip

jni committed:
skip failing CI config

jni committed:
Mark the right test as xfail 🤦

Czaki commented on napari/components/camera.py:
```diff
@@ -1,18 +1,42 @@
+    DepthAxisOrientation.TOWARDS,
+    VerticalAxisOrientation.DOWN,
+    HorizontalAxisOrientation.RIGHT,
+)
```
```suggestion
DEFAULT_ORIENTATION_TYPED = (
    DepthAxisOrientation.TOWARDS,
    VerticalAxisOrientation.DOWN,
    HorizontalAxisOrientation.RIGHT,
)
DEFAULT_ORIENTATION = tuple(map(str, DEFAULT_ORIENTATION_TYPED))
```

Make a single point of edit

Czaki commented on napari/components/camera.py:
```diff
@@ -213,6 +242,18 @@ def calculate_nd_up_direction(
+    def orientation2d(
+        self, value: tuple[VerticalAxisOrientation, HorizontalAxisOrientation]
+    ) -> None:
+        self.orientation = (self.orientation[0],) + value
```
We use `StrEnum` why do not allow passing by name? Maybe fix type annotation?

jni replied:
Oh it's just me being bad at typing. Thanks for the hint! Will update.

jni replied:
Hmmm, @Czaki now that I added `| str` to the types in the method parameter, what's the best way to tell mypy that pydantic validation + StrEnum coercion means this assignment is fine?

```
napari/components/camera.py:258: error: Incompatible types in assignment
(expression has type
"tuple[DepthAxisOrientation, VerticalAxisOrientation | str, HorizontalAxisOrientation | str]",
variable has type
"tuple[DepthAxisOrientation, VerticalAxisOrientation, HorizontalAxisOrientation]")
 [assignment]
            self.orientation = (self.orientation[0],) + value
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

jni replied:
(failure is here: https://github.com/napari/napari/actions/runs/13654349055/job/38169835992?pr=7663#step:5:16)

jni replied:
I've just ignored that line for now — at least the public API is correctly typed now. Though I guess anyone using the camera model and running their code through mypy will run into the same error...

Czaki commented on napari/_vispy/camera.py:
```diff
@@ -176,6 +187,26 @@ def _on_center_change(self):
+        self._2D_camera.flip = tuple(
+            int(ori != default_ori)
+            for ori, default_ori in zip(
+                orientation_xyz, VISPY_DEFAULT_ORIENTATION_2D, strict=False
```
Why strict False? As I read code I expect that both are of lenght 3.

jni replied:
Ha, well spotted. The reason is that I didn't have strict= at all, and ruff added it, so I left it as ruff did. I'll update.

Czaki commented on napari/_vispy/camera.py:
```diff
@@ -3,6 +3,15 @@
+# is passed from napari to Vispy, usually with a transposition. In the camera
+# models, this means that the order of these orientations appear in the
+# opposite order to that in napari.components.Camera.
+VISPY_DEFAULT_ORIENTATION_2D = ('right', 'up', 'towards')  # xyz
+VISPY_DEFAULT_ORIENTATION_3D = ('right', 'down', 'away')  # xyz
+
```
Why we use different orientations in 2D and 3D, I do not feel like comment is explaining this.

jni replied:
We don't use it, Vispy uses it. I don't know why, it's just something I observed.

jni replied:
Added a note in the comment.

Czaki commented on napari/_vispy/_tests/test_vispy_camera.py:
```diff
@@ -138,3 +143,75 @@ def test_camera_model_update_from_vispy_3D(make_napari_viewer):
+
+    viewer.camera.perspective = 60
+    viewer.camera.orientation = ('away', 'down', 'right')
+    sshot_away = viewer.screenshot(canvas_only=True, flash=False)[..., 0]
```
You need to process qt events before making a screenshot.

jni replied:
What's the incantation to do this? And should we add an event processing call to screenshot? I have run into this too many times… 😭

Czaki replied:
> And should we add an event processing call to screenshot? I

This is required only for testing. 

The invocation is `QApplication.processEvents()`

jni committed:
Use single source of truth for default orientation

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

jni committed:
Process events in tests before taking screenshot

jni committed:
Use zip(..., strict=True) where possible

jni committed:
Fix typing on orientation2d setter

jni committed:
Add note about vispy default orientations

jni committed:
Remove manual setting of camera flips

They get updated later with self._on_ndisplay_change() ->
self._on_orientation_change()

jni committed:
Ignore typing issue with assignment

jni committed:
Revert "Process events in tests before taking screenshot"

This reverts commit 0fc4b70852800d4faf24d20db8e438c7c4d251c3.

jni committed:
Make xfail non-strict since sometimes it can pass

Czaki commented on napari/components/camera.py:
```diff
@@ -1,18 +1,42 @@
 import warnings
+from enum import auto
 from typing import TYPE_CHECKING, Optional
```
```suggestion
from typing import TYPE_CHECKING, Literal, Optional
```

Czaki commented on napari/components/camera.py:
```diff
@@ -1,18 +1,42 @@
     import numpy.typing as npt
 
 
+class VerticalAxisOrientation(StringEnum):
+    UP = auto()
+    DOWN = auto()
+
+
+class HorizontalAxisOrientation(StringEnum):
+    LEFT = auto()
+    RIGHT = auto()
+
+
+class DepthAxisOrientation(StringEnum):
+    AWAY = auto()
+    TOWARDS = auto()
+
```
```suggestion
class VerticalAxisOrientation(StringEnum):
    UP = auto()
    DOWN = auto()

VerticalAxisOrientationStr = Literal['up'] | Literal['down']


class HorizontalAxisOrientation(StringEnum):
    LEFT = auto()
    RIGHT = auto()
    
HorizontalAxisOrientationStr = Literal['left'] | Literal['right']

class DepthAxisOrientation(StringEnum):
    AWAY = auto()
    TOWARDS = auto()
    
HorizontalAxisOrientationStr = Literal['away'] | Literal['torwards']

```

Czaki commented on napari/components/camera.py:
```diff
@@ -213,6 +242,23 @@ def calculate_nd_up_direction(
+    def orientation2d(
+        self,
+        value: tuple[
+            VerticalAxisOrientation | str, HorizontalAxisOrientation | str
```
```suggestion
            VerticalAxisOrientation | VerticalAxisOrientationStr, HorizontalAxisOrientation | HorizontalAxisOrientationStr
```

Czaki commented on napari/components/camera.py:
```diff
@@ -213,6 +242,23 @@ def calculate_nd_up_direction(
+    ) -> None:
+        # we ignore typing because Pydantic + StrEnum can correctly coerce
+        # str values to their corresponding StrEnum values in the model
+        self.orientation = (self.orientation[0],) + value  # type:ignore[assignment]
```
```suggestion
        self.orientation = (self.orientation[0], VerticalAxisOrientation(value[0]), HorizontalAxisOrientation(value[1]))
```

Czaki reviewed (dismissed):
Please type it correctly

jni committed:
Proper typing of StrEnum strings

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jni committed:
Rearrange and fix strenum string types
